rfcs/005-metadata-rework.md

```diff
@@ -0,0 +1,562 @@
+fields (see [#815](https://github.com/tweag/nickel/pull/815)). It seems natural
+to store them as metadata (same as for default values). But then, record
+operations such as `record.fields` need to decide "is this field is optional and
+empty", and a linear time operation suddenly becomes arbitrary long as it might
+imply to evaluate each and every field.
+
+#### Nested metadata
+
+The second issue With metadata anywhere is the semantics of nested metavalues.
+Currently, when fetching, we stop at the first metadata we encounter. But they
+may very well be nested. The current implementation is oblivious to potential
+inner metadata:
+
+```text
+nickel>let base = {
+  foo | doc "Foo"
+      | Str
+      | default = "bar"
+}
+nickel>:query base.foo
+• contract: Str
+• default: "bar"
+• documentation: Foo
+
+nickel>let base_hidden = {
+  foo = ((("bar" | doc "Foo") | default) | Str)
+}
+nickel>:query base_hidden.foo
+• contract: Str
+• value: "bar"
+
+```
+
+One solution could be to define a flattening operation (pedantically, a monoid
+operation, or a monad if metadata are rather seen as a container over a value):
+`flatten : (Metadata, Metadata) -> Metadata`.
+
+While fetching metadata can take an arbitrary time to complete _in principle_,
+it's currently quick for the happy path, when metadata are located e.g. directly
+on record fields. Now, if we now want to correctly handle nested metadata, we
+can't stop evaluation as soon as we encounter metadata anymore, because some
+other metadata could be lying underneath. We now have to _systematically
+evaluate an expression down to a value_, collecting and flattening metadata
+along the way.
+
+## Proposal
+
+In a configuration language such as Nickel, the fundamental data structure is
+the record, and the fundamental piece of data is the record field. In
+retrospect, fields seem to be the right level of granularity for metadata, and
+where they are most often attached in practice.
+
+While metadata that can appear anywhere are in theory more flexible, the only
+example making use of such capability of metadata to be attached to other values
+than record fields is the example where metadata are propagated through an array
+and another record field. This example is contrived, and doesn't seem to
+correspond to a real use-case in a convincing way. In the NixOS module system,
+for example, metadata are attached to options only.
+
+Field-based metadata would only be affected during merging, and combined eagerly
+when that happens: no later arbitrary evaluation, no nesting, just static data.
+Allowing them anywhere doesn't bring clear practical benefits, whereas it comes
+with substantial drawbacks, exposed in the previous sections.
+
+### Specification
+
+The proposed change comes in to aspects:
+
+1. Metadata can only be written next to record fields, and are stored only there
+   in the AST. Their semantics with respect to merging is left unchanged, except
```

Comment:
As I've told you in off-channel conversations, I think they could be attached to let bindings as well. Not sure that it's quite as useful though. So maybe we don't have to think about it quite now.

Reply:
It could make sense indeed, to attaching documentation for example. But I think it may bring the same issue discussed here if we allow for example default values. I would vouch for let's implement this for record fields only, and consider later an extension if this seems to be really needed.